rust/src/daba/mod.rs

```diff
@@ -0,0 +1,151 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::collections::VecDeque;
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+pub struct DABA<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    // ith oldest value in FIFO order stored at vi = vals[i]
+    vals: VecDeque<Value>,
+    aggs: VecDeque<Value>,
+    // 0 ≤ l ≤ r ≤ a ≤ b ≤ aggs.len()
+    l: usize, // Left,  ∀p ∈ l...r−1 : aggs[p] = vals[p] ⊕ ... ⊕ vals[r−1]
+    r: usize, // Right, ∀p ∈ r...a−1 : aggs[p] = vals[R] ⊕ ... ⊕ vals[p]
+    a: usize, // Accum, ∀p ∈ a...b−1 : aggs[p] = vals[p] ⊕ ... ⊕ vals[b−1]
+    b: usize, // Back,  ∀p ∈ b...e−1 : aggs[p] = vals[B] ⊕ ... ⊕ vals[p]
+    op: PhantomData<BinOp>,
+}
+
+impl<Value, BinOp> FifoWindow<Value, BinOp> for DABA<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new() -> Self {
+        Self {
+            vals: VecDeque::new(),
+            aggs: VecDeque::new(),
+            l: 0,
+            r: 0,
+            a: 0,
+            b: 0,
+            op: PhantomData,
+        }
+    }
+    fn push(&mut self, v: Value) {
+        self.aggs.push_back(self.agg_b().operate(&v));
+        self.vals.push_back(v);
+        self.fixup();
+    }
+    fn pop(&mut self) {
+        if self.vals.pop_front().is_some() {
+            self.aggs.pop_front();
+            self.l -= 1;
```

Comment:
I'm not sure decrementing is the right thing here - it may be, but it should be carefully considered.

You've implemented `l`, `r`, `a` and `b` as indices, which is something we're just now grappling with: we implemented them as iterators, but our presentation in the 2017 paper was kinda loose with are they iterators/pointers or indices. When they're indices, that means `f==0` and `e==vals.len()` is always true, and don't need to explicitly represent them. I think you've correctly done that. But I'm unsure about this decrementing - you're following our published algorithms for `fixup()`, which did not assume decrementing on eviction. It's also possible there's a subtly I'm missing with the semantics of `VecDeque`.

Reply:
In Rust, it's possible to overload the [indexing](https://doc.rust-lang.org/std/ops/trait.Index.html) operator. The indexing for `VecDeque` is an alias for [get](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.get) which indexes from the front (including offset). I decrement all indices when removing the front element to shift them left.

Reply:
Yes, now I see what you mean. Calling `pop_front()` on a `VecDeque` effectively shifts the entire deque to the left, so if you had an index for location `i`, accessing `q[i]` before and after the pop will yield different values. You're decrementing `i` so that it continues to point at the same value in the deque.

I was wondering if it would be easier to just use an iterator, and I don't think it is. From what I can tell, there's no easy way to "decrement" a standard Rust iterator as is required in the shrink case. So, in effect, you have to do your own iterator management.

Reply:
I think what we want is a linked list of arrays with ["cursors"](https://doc.rust-lang.org/beta/std/collections/linked_list/struct.Cursor.html).

Reply:
Yes, the concept of a cursor is essentially what C++ style iterators are, and how we ended up presenting the algorithm. The chunked array queue is a linked list, it just has many elements in each node.